### PR TITLE
[camsoda] Add extractor

### PR DESCRIPTION
**Pull request by zulaport**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Adds an extractor for the adult webcam site, [Camsoda](https://camsoda.com). Manual test:

```
▶ python -m yt_dlp https://www.camsoda.com/lizzhopf
[Camsoda] lizzhopf: Downloading webpage
[Camsoda] lizzhopf: Downloading JSON metadata
[Camsoda] lizzhopf: Downloading m3u8 information
[info] lizzhopf: Downloading 1 format(s): hls-1
[download] Destination: lizzhopf 2022-11-06 13_22 [lizzhopf].mp4
```

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/camsoda.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            raise ExtractorError(self._ROOM_OFFLINE, expected=True)
+
+        token = traverse_obj(data, ('token'), expected_type=str)
+
+        for server in traverse_obj(data, ('edge_servers'), expected_type=list):
+            formats = self._extract_m3u8_formats(
+                f'https://{server}/{stream_name}_v1/index.m3u8?token={token}',
+                video_id, ext='mp4', m3u8_id='hls', fatal=False, live=True)
+            if formats:
+                break
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:51]:
```suggestion
        if data.get('private_servers'):
            raise ExtractorError(self._ROOM_PRIVATE, expected=True)
        elif not data.get('stream_name'):
            raise ExtractorError(self._ROOM_OFFLINE, expected=True)

        token = traverse_obj(data, 'token', expected_type=str)
        
        for server in traverse_obj(data, ('edge_servers', ...)):
            formats = self._extract_m3u8_formats(
                f'https://{server}/{stream_name}_v1/index.m3u8?token={token}',
                video_id, ext='mp4', m3u8_id='hls', fatal=False, live=True)
            if formats:
                break
```

File: yt_dlp/extractor/camsoda.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': video_id,
+            'description': self._html_search_meta('description', webpage, default=None),
+            'is_live': True,
+            'formats': formats,
+            # Camsoda does not declare the RTA meta-tag
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:61]:
```suggestion
```

File: yt_dlp/extractor/camsoda.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        if not data:
+            raise ExtractorError('Unable to find configuration for stream.')
+
+        private_servers = traverse_obj(data, ('private_servers'), expected_type=list)
+        if private_servers:
+            raise ExtractorError(self._ROOM_PRIVATE, expected=True)
+
+        stream_name = traverse_obj(data, ('stream_name'), expected_type=str)
+        if not stream_name:
+            raise ExtractorError(self._ROOM_OFFLINE, expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:42]: shouldnt these checks be done only if there is no format?

File: yt_dlp/extractor/camsoda.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        },
+        'skip': 'Room is offline',
+    }]
+
+    _ROOM_OFFLINE = 'Model is offline.'
+    _ROOM_PRIVATE = 'Model is in private show.'
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        user_id = 'guest_%u' % random.randrange(10000, 99999)
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:27]: fstring

File: yt_dlp/extractor/camsoda.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            'title': 'lizzhopf (lizzhopf) Nude on Cam. Free Live Sex Chat Room - CamSoda',
+            'description': str,
+            'is_live': True,
+            'age_limit': 18,
+        },
+        'skip': 'Room is offline',
+    }]
+
+    _ROOM_OFFLINE = 'Model is offline.'
+    _ROOM_PRIVATE = 'Model is in private show.'
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:23]: single use vars

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/camsoda.py (hunk 3 above)
Comment (zulaport) [reply yt_dlp/extractor/camsoda.py:42]: If the room is offline, there server data isn't available in the response, so I thought checking these conditions first would make more sense.

Online response:

```json
{
    "aspect": 16,
    "edge_servers": [
        "edge54-ord.livemediahost.com",
        "edge55-ord.livemediahost.com"
    ],
    "ffmpeg_server": "ded4589",
    "height": 1080,
    "status": 1,
    "stream_name": "cam_obs/catydance-flu",
    "token": "[REDACTED]",
    "width": 1920
}
```

Offline response:

```json
{
    "aspect": 16,
    "edge_servers": [],
    "ffmpeg_server": null,
    "height": 1080,
    "status": 1,
    "stream_name": "",
    "token": "[REDACTED]",
    "width": 1920
}
```

Private show response:

```json
{
    "aspect": 16,
    "edge_servers": [],
    "ffmpeg_server": "ded3441",
    "height": 540,
    "status": 1,
    "stream_name": "1280x720/agataruiz-flu-ingest7-bog",
    "token": "[REDACTED]",
    "width": 960,
    "private_servers": [
        "ingest7-bog.livemediahost.com"
    ]
}
```

File: yt_dlp/extractor/camsoda.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                f'https://{server}/{stream_name}_v1/index.m3u8?token={token}',
+                video_id, ext='mp4', m3u8_id='hls', fatal=False, live=True)
+            if formats:
+                break
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': video_id,
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:52]: why is title=id? This also doesnt match test

File: yt_dlp/extractor/camsoda.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        user_id = f'guest_{random.randrange(10000, 99999)}'
+        webpage = self._download_webpage(url, video_id, headers=self.geo_verification_headers())
+
+        data = self._download_json(
+            f'https://camsoda.com/api/v1/video/vtoken/{video_id}?username={user_id}',
+            video_id, headers=self.geo_verification_headers())
+        if not data:
+            raise ExtractorError('Unable to find configuration for stream.')
+
+        if data.get('private_servers'):
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:33]:
```suggestion
        elif data.get('private_servers'):
```

File: yt_dlp/extractor/camsoda.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
+import random
+
+from .common import InfoExtractor
+from ..utils import ExtractorError, traverse_obj
+
+
+class CamsodaIE(InfoExtractor):
+    _VALID_URL = r'https?://www\.camsoda\.com/(?P<id>[0-9A-Za-z-]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:8]:
```suggestion
    _VALID_URL = r'https?://www\.camsoda\.com/(?P<id>[\w-]+)'
```

File: yt_dlp/extractor/camsoda.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        user_id = f'guest_{random.randrange(10000, 99999)}'
+        webpage = self._download_webpage(url, video_id, headers=self.geo_verification_headers())
+
+        data = self._download_json(
+            f'https://camsoda.com/api/v1/video/vtoken/{video_id}?username={user_id}',
+            video_id, headers=self.geo_verification_headers())
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:29]:
```suggestion
        webpage = self._download_webpage(url, video_id, headers=self.geo_verification_headers())

        data = self._download_json(
            f'https://camsoda.com/api/v1/video/vtoken/{video_id}', video_id,
            query={'username': f'guest_{random.randrange(10000, 99999)}',
            headers=self.geo_verification_headers())
```

File: yt_dlp/extractor/camsoda.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            'is_live': True,
+            'age_limit': 18,
+        },
+        'skip': 'Room is offline',
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        user_id = f'guest_{random.randrange(10000, 99999)}'
+        webpage = self._download_webpage(url, video_id, headers=self.geo_verification_headers())
```
Comment (pukkandan) [inline yt_dlp/extractor/camsoda.py:25]: is it worth downloading webpage just for description?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/camsoda.py (hunk 10 above)
Comment (zulaport) [reply yt_dlp/extractor/camsoda.py:24]: What about to also extract the title to address your other comment?

File: yt_dlp/extractor/camsoda.py (hunk 6 above)
Comment (zulaport) [reply yt_dlp/extractor/camsoda.py]: This is the same approach in the stripchat extractor, but I see other extractors use `self._html_extract_title` so I did that here too now.

File: yt_dlp/extractor/camsoda.py (hunk 10 above)
Comment (pukkandan) [reply yt_dlp/extractor/camsoda.py:24]: I was just asking. I don't use the site - it's your discretion whether the metadata is worth the additional request

Comment (zulaport) [reply yt_dlp/extractor/camsoda.py:24]: I don't use this metadata either, but if this metadata is useful to someone else, it does look like retrieving the page is the only way to get it so I think it makes sense.

**Pull request merged**